File: cors/resources/cors-makeheader.php
<?php

$origin = isset($_GET['origin']) ? $_GET['origin'] : $_SERVER['HTTP_ORIGIN'];

if ($origin != 'none')
    header("Access-Control-Allow-Origin: $origin");
if (isset($_GET['origin2']))
    header("Access-Control-Allow-Origin: {$_GET['origin2']}", false);

/* Preflight */
if (isset($_GET['headers']))
    header("Access-Control-Allow-Headers: {$_GET['headers']}");
if (isset($_GET['credentials']))
    header("Access-Control-Allow-Credentials: {$_GET['credentials']}");
if (isset($_GET['methods']))
    header("Access-Control-Allow-Methods: {$_GET['methods']}");

$code = isset($_GET['code']) ? intval($_GET['code']) : null;
if ($_SERVER['REQUEST_METHOD'] == 'OPTIONS' and isset($_GET['preflight']))
    $code = intval($_GET['preflight']);

if (isset($_GET['location']))
{
    if ($code === null)
    	$code = 302;

    if ($code < 400 and $code > 299)
    {
        header("Location: {$_GET['location']}", true, $code);
        die("Redirecting");
    }
}

foreach ($_SERVER as $name => $value)
{
    if (substr($name, 0, 5) == 'HTTP_')
    {
        $name = strtolower(str_replace('_', '-', substr($name, 5)));
        $headers[$name] = $value;
    } else if ($name == "CONTENT_TYPE") {
        $headers["content-type"] = $value;
    } else if ($name == "CONTENT_LENGTH") {
        $headers["content-length"] = $value;
    }
}

$headers['get_value'] = isset($_GET['get_value']) ? $_GET['get_value'] : '';

if ($code)
    header("HTTP/1.1 {$code} StatusText");

echo json_encode( $headers );

File: cors/resources/cors-makeheader.py
import json

def main(request, response):
    origin = request.GET.first(b"origin", request.headers.get(b'origin'))

    if b"check" in request.GET:
        token = request.GET.first(b"token")
        value = request.server.stash.take(token)
        if value is not None:
            if request.GET.first(b"check", None) == b"keep":
                request.server.stash.put(token, value)
            body = u"1"
        else:
            body = u"0"
        return [(b"Content-Type", b"text/plain")], body


    if origin != b'none':
        response.headers.set(b"Access-Control-Allow-Origin", origin)
    if b'origin2' in request.GET:
        response.headers.append(b"Access-Control-Allow-Origin", request.GET.first(b'origin2'))

    #Preflight
    if b'headers' in request.GET:
        response.headers.set(b"Access-Control-Allow-Headers", request.GET.first(b'headers'))
    if b'credentials' in request.GET:
        response.headers.set(b"Access-Control-Allow-Credentials", request.GET.first(b'credentials'))
    if b'methods' in request.GET:
        response.headers.set(b"Access-Control-Allow-Methods", request.GET.first(b'methods'))

    code_raw = request.GET.first(b'code', None)
    if code_raw:
        code = int(code_raw)
    else:
        code = None
    if request.method == u'OPTIONS':
        #Override the response code if we're in a preflight and it's asked
        if b'preflight' in request.GET:
            code = int(request.GET.first(b'preflight'))

        #Log that the preflight actually happened if we have an ident
        if b'token' in request.GET:
            request.server.stash.put(request.GET[b'token'], True)

    if b'location' in request.GET:
        if code is None:
            code = 302

        if code >= 300 and code < 400:
            response.headers.set(b"Location", request.GET.first(b'location'))

    headers = {}
    for name, values in request.headers.items():
        if len(values) == 1:
            headers[name.decode("iso-8859-1")] = values[0].decode("iso-8859-1")
        else:
            #I have no idea, really
            headers[name] = values

    headers[b'get_value'.decode("iso-8859-1")] = request.GET.first(b'get_value', b'').decode("iso-8859-1")

    body = json.dumps(headers)

    if code:
        return (code, u"StatusText"), [], body
    else:
        return body
